### PR TITLE
feat: Client-side rate limiting

### PR DESCRIPTION
**Pull request by jbelkins**
## Issue \#
https://github.com/awslabs/aws-sdk-swift/issues/338

## Description of changes
Adds the "adaptive" mode of retry described in Retry Behavior 2.0, by adding a client-side rate limiter that can be optionally enabled when configuring an AWS service client.

also: Cleaned up tests a little bit.

## Scope
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: Sources/ClientRuntime/Retries/DefaultRetryStrategy/ClientSideRateLimiter.swift [hunk 1]
```diff
@@ -0,0 +1,155 @@
...
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import struct Foundation.TimeInterval
+import struct Foundation.Date
+import func Foundation.pow
+
+actor ClientSideRateLimiter {
```
Comment (jbelkins) [inline Sources/ClientRuntime/Retries/DefaultRetryStrategy/ClientSideRateLimiter.swift:12]: This type follows the Retry Behavior 2.0 specification very closely.

File: Sources/ClientRuntime/Retries/DefaultRetryStrategy/DefaultRetryToken.swift [hunk 2]
```diff
@@ -11,15 +11,12 @@ import struct Foundation.TimeInterval
...
 /// until the request succeeds or fails after zero or more retries.
-public class DefaultRetryToken: RetryToken {
+public final class DefaultRetryToken: RetryToken {
 
     /// The number of retry attempts that have been made using this token.
     /// Defaults to zero at the initial attempt, goes up by one for each subsequent attempt.
     public internal(set) var retryCount: Int = 0
 
-    /// The delay, in seconds, to the next retry.
-    public internal(set) var delay: TimeInterval?
```
Comment (jbelkins) [inline Sources/ClientRuntime/Retries/DefaultRetryStrategy/DefaultRetryToken.swift:21]: This delay param on token is not used in the Swift implementation, hence it is removed.

File: Sources/ClientRuntime/Retries/DefaultRetryStrategy/RetryQuota.swift [hunk 3]
```diff
@@ -59,4 +83,12 @@ final actor RetryQuota {
         availableCapacity += capacityAmount ?? Self.noRetryIncrement
         availableCapacity = min(availableCapacity, maxCapacity)
     }
+
```
Comment (jbelkins) [inline Sources/ClientRuntime/Retries/DefaultRetryStrategy/RetryQuota.swift:86]: These 2 methods below expose the new rate limiting functions.

File: Sources/ClientRuntime/Retries/RetryToken.swift [hunk 4]
```diff
@@ -14,7 +14,4 @@ public protocol RetryToken: AnyObject {
 
     /// The number of retries (i.e. NOT including the initial attempt) that this token has made.
     var retryCount: Int { get }
-
-    /// The delay for this request (TODO: not used, maybe get rid of this?)
-    var delay: TimeInterval? { get }
```
Comment (jbelkins) [inline Sources/ClientRuntime/Retries/RetryToken.swift:19]: Unused property.

File: Tests/ClientRuntimeTests/Retry/DefaultRetryStrategy/ClientSideRateLimiterTests.swift [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+// All Rights Reserved.
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+import Foundation
+import XCTest
+@testable import ClientRuntime
+
+final class ClientSideRateLimiterTests: XCTestCase {
```
Comment (jbelkins) [inline Tests/ClientRuntimeTests/Retry/DefaultRetryStrategy/ClientSideRateLimiterTests.swift:12]: These tests are straight out of the Retry Behavior 2.0 spec.

**Review by waahm7 — APPROVED** [review]

**Pull request merged**